test/test_mbarrier.cpp

```diff
@@ -0,0 +1,125 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <gmock/gmock-matchers.h>
+#include <gtest/gtest.h>
+
+#include <executor.h>
+#include <fusion.h>
+#include <ops/all_ops.h>
+#include <test/utils.h>
+#include <test/validator.h>
+
+namespace nvfuser {
+
+class MBarrierTest : public NVFuserTest {};
+
+TEST_F(MBarrierTest, Simple) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  TensorView* tv0 = makeContigConcreteTensor({32, 32});
+  auto tv1 = set(tv0);
+  auto tv2 = set(tv1);
+  fusion.addInput(tv0);
+  fusion.addOutput(tv2);
+
+  tv1->setMemoryType(MemoryType::Shared);
+  tv1->axis(0)->parallelize(ParallelType::TIDx);
+  tv1->axis(1)->parallelize(ParallelType::TIDy);
+
+  tv2->axis(0)->parallelize(ParallelType::TIDy);
+  tv2->axis(1)->parallelize(ParallelType::TIDx);
+
+  FusionExecutor fe;
+
+  fe.registerPostLoweringHook([](kir::Kernel* kernel) {
```

Comment:
Kernel after modification:
```CUDA
__global__ void kernel1(Tensor<float, 2, 2> T0, Tensor<float, 2, 2> T2) {
  alignas(16) extern __shared__ char array[];
  const unsigned smem_offset = 0;
  nvfuser_index_t i0;
  i0 = ((nvfuser_index_t)threadIdx.y) + (32 * ((nvfuser_index_t)threadIdx.x));
  nvfuser_index_t i1;
  i1 = ((nvfuser_index_t)threadIdx.x) + (32 * ((nvfuser_index_t)threadIdx.y));
  float* T1 = reinterpret_cast<float*>(array + smem_offset + 0);
  uint64_t* T3 = reinterpret_cast<uint64_t*>(array + smem_offset + 4096);
  mbarrier::init(toSmem(T3), 1024);
  T1[i0]
     = T0[i0];
  uint64_t i2;
  i2 = mbarrier::arrive(toSmem(T3));
  mbarrier::wait(toSmem(T3), i2);
  T2[i1]
     = T1[i1];
  mbarrier::inval(toSmem(T3));
}
```

Reply:
Thanks for this. This is really a nice way of testing initial build-out features